### PR TITLE
Context item addons (part II)

### PR DESCRIPTION
**Pull request by tamland**
Since @Fice is unavailable/busy I'm opening a new PR where I've tried to fix/finish up the last few bits of #1654.

It's done now. Needs reviewing and testing please:)

As mentioned, category addons are not included here as the feature is not finished. For detailed list of changes see https://github.com/tamland/xbmc/commits/context_addons_fixup

### CODE REVIEW

**Comment by tamland** [comment]
@Montellese Care to do a review? If the other one was ok.. the points you made should be resolved and changes should be minor. Btw. this conflicts with some of the changes in the webserver PR. Perhaps you want to merge parts of that first?

**Comment by tamland** [comment]
@topfs2 Since you assigned yourself to the other PR, is there anything you would like to add:)?

**Comment by tamland** [comment]
thanks for detailed review @Montellese!

**Comment by Montellese** [comment]
@tamland: Haven't had time yet to look at the actual implementation but hope to do so later today.

**Comment by tamland** [comment]
great:)

**Comment by Montellese** [comment]
Two questions since I didn't follow the discussion in the original PR:
* Is the `parent` feature already used?
* Does every context item require a separate extension point?

**Comment by tamland** [comment]
Man, felt like I already went over this so many times fixed so many issues..

To answer your questions: the parent feature is not the same as the original PR, as it needs more work. You have only two options here: no parent or 'xbmc.manage' to add it in the manage menu. Can drop this if you don't think that will be useful. I just though it might.

And you can only do *one* context item per addon here, because the whole addon manger is designed around max one point per type. I'm thinking as a possible future extension that multiple items (and hierarchy) can be defined as an xml structure in the <extension> tag. That's why I left that open.

**Comment by Montellese** [comment]
Hehe only a few minor comments, nothing serious.

If we ever want to support multiple context items per addon using the current attributes is probably not the best approach as the attributes are limited to the extension point element. For that case ideally the CContextItemAddon class would also provide the number of items it provides so that the IDs can be properly assigned in the CContextMenuManager. But that can obviously be done later.

**Comment by tamland** [comment]
Rebased and changed IDs to `kodi.context.item` and `kodi.core.manage` since that's apparently what we're going for now.

**Comment by tamland** [comment]
Need project sync. @mkortstiege  you're the go-to guy for that nowadays?

**Comment by mkortstiege** [comment]
Looks like it ;) Here we go .. https://github.com/mkortstiege/xbmc/commit/7a0765af2a7efd1b6cf15c8fac7824cb103b6198

**Comment by Montellese** [comment]
Do you intend to keep using attributes for `label` and `visible`? I simply think it's not future proof and will make backwards compatibility harder if we ever want to support multiple context menu items per extension point.

**Comment by tamland** [comment]
@Montellese Well, I though it would be the opposite. That if you add `label` and `visible` tags you have to commit to that structure making backwards compatibility harder. E.g. if you find out later that everything should be inside `<category><item>` tags, you have two different xml structures you have to support. I only kept the attributes because of that; tags are certainly better and easier to read/write as visibility expressions can get very long.

**Comment by Montellese** [comment]
Well I just figured that an extension point tag can only take one `label` and one `visible` attribute so there's no way you'll ever be able to extend that approach to support multiple items. If you use something like
```xml
<item label="foo" visible="bar" />
```
or (more in line with the skinning syntax but also more verbose)
```xml
<item>
  <label>foo</label>
  <visible>bar</visible>
</item>
```
you can easily define multiple items per extension point.

**Comment by tamland** [comment]
Never intended attributes to be *extended* it, but killed and replace it with tags once we figured out how to do categories and multiple items.

**Comment by tamland** [comment]
Your example is pretty much exactly what I had in mind for the future, but it's missing groups, ways to extend existing groups, and what to do the the `library` attribute which have to be changed to allow multiple scripts.

**Comment by tamland** [comment]
Ok, changed it to top level tags and made that whole structure somewhat stricter (no point leaving room for ambiguity)

**Comment by tamland** [comment]
@Montellese Updated schema, but I have no idea if that's the correct way to define it

**Comment by tamland** [comment]
jenkins build this please

**Comment by tamland** [comment]
jenkins build this please

**Comment by tamland** [comment]
No idea what that osx link error means. Are you sure the project files are correct here @mkortstiege?

**Comment by mkortstiege** [comment]
@tamland will check.

**Comment by mkortstiege** [comment]
@tamland please pick https://github.com/mkortstiege/xbmc/commit/0a8d21cfc1b6ae44b89e3a09beceed1ed4aa1a4f.

**Comment by tamland** [comment]
@Montellese Can you do a proper VS sync? Can't battle with that any longer. (preferably one that easily rebasable if you merge first:))

**Comment by tamland** [comment]
jenkins build this please

**Comment by Montellese** [comment]
@tamland: VS project files always conflict. It's a PITA. I don't understand why git can't properly merge XML files which have such an obvious and easy structure.

File: addons/xbmc.python/contextitem.xsd [hunk 1]
```diff
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE schema PUBLIC "-//W3C//DTD XMLSCHEMA 200102//EN" "http://www.w3.org/2001/XMLSchema.dtd">
+<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema">
+  <xs:element name="extension">
+    <xs:complexType>
+      <xs:attribute name="point" type="xs:string" use="required"/>
+      <xs:attribute name="id" type="simpleIdentifier"/>
+      <xs:attribute name="name" type="xs:string"/>
+      <xs:attribute name="library" type="xs:string" use="required"/>
+      <xs:sequence>
```
Comment (Montellese) [inline addons/xbmc.python/contextitem.xsd]: IIRC `<xs:sequence>` needs to be placed before `<xs:attribute>`.

**Comment by Montellese** [comment]
@tamland: Could you provide an example extension point definition? Is the current layout now extensible to multiple items if we want?

**Comment by tamland** [comment]
jenkins build this please

**Comment by FernetMenta** [comment]
it is not a good idea doing project file updates for vs or xcode in a separate commit. when it comes to bisect we are in trouble because not every SHA builds correctly.

**Comment by tamland** [comment]
Makes sense. Might want to spread the word because I see a lot of other people doing it in separate commits too.

**Pull request merged**